tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc

```diff
@@ -345,6 +345,60 @@ static LogicalResult Verify(CustomCallOp op) {
   return verify_types_and_layouts(result_types, result_layouts, "result");
 }
 
+//===----------------------------------------------------------------------===//
+// DotOp
+//===----------------------------------------------------------------------===//
+namespace {
+bool dimCompatible(int64_t a, int64_t b) {
+  return ShapedType::kDynamicSize == a || ShapedType::kDynamicSize == b ||
+         a == b;
+}
+
+ShapedType inferDotReturnType(ShapedType lhs, ShapedType rhs) {
+  auto element_type = lhs.getElementType();
+  if (!lhs.hasRank() || !rhs.hasRank()) {
+    return UnrankedTensorType::get(element_type);
+  }
+
+  // vector dot vector
+  if (1 == lhs.getRank() && 1 == rhs.getRank() &&
+      dimCompatible(lhs.getDimSize(0), rhs.getDimSize(0))) {
+    return RankedTensorType::get({}, element_type);
+  }
+  // matrix dot vector
+  if (2 == lhs.getRank() && 1 == rhs.getRank() &&
+      dimCompatible(lhs.getDimSize(1), rhs.getDimSize(0))) {
+    return RankedTensorType::get({lhs.getDimSize(0)}, element_type);
+  }
+  // matrix dot matrix
+  if (2 == lhs.getRank() && 2 == rhs.getRank() &&
+      dimCompatible(lhs.getDimSize(1), rhs.getDimSize(0))) {
+    int64_t shape[2] = {lhs.getDimSize(0), rhs.getDimSize(1)};
+    return RankedTensorType::get(shape, element_type);
+  }
+  return {};
+}
+} // namespace
+
+static LogicalResult Verify(DotOp op) {
+  auto lhs_type = op.lhs().getType().dyn_cast<ShapedType>();
+  auto rhs_type = op.rhs().getType().dyn_cast<ShapedType>();
+  auto result_type = op.getType().dyn_cast<ShapedType>();
+  if (!lhs_type || !rhs_type || !result_type) {
+    return op.emitError() << "Unexpected operands or result type.";
+  }
+  auto expect_return_type = inferDotReturnType(lhs_type, rhs_type);
+  if (!expect_return_type) {
+    return op.emitError() << "Unexpected operands type.";
+  }
+  if (result_type.hasRank() && expect_return_type.hasRank()) {
+    if (result_type.getShape() != expect_return_type.getShape()) {
+      return op.emitError() << "Unexpected result type.";
```

Comment:
Something that helps for every error message is to print more info about the context, for example here:
```
      return op.emitError() << "Unexpected result type: " << result_type
                            << " vs " << expect_return_type;
```